### PR TITLE
BUG: convolve may yield inconsistent dtypes with method changed

### PR DESCRIPTION
**Pull request by stsievert**
Addresses #7193.

In this I cast the output as the value returned by `np.return_types` (thanks to @e-q's suggestion).

I also implement a test that ensures all 1D dtypes are the same when convolving with all possible types (two arrays are of different types). In #5608 I only tested to make sure that the dtypes matched when convolving two arrays *of the same type*.

~~Let me do some more verification before merge.~~

### CODE REVIEW

File: scipy/signal/tests/test_signaltools.py [hunk 1]
```diff
@@ -156,40 +156,55 @@ def test_invalid_shapes(self):
...
+                    x += 1j * np.random.rand(*x.shape).astype(dtype)
+                    h += 1j * np.random.rand(*h.shape).astype(dtype)
+
+                results = {key: convolve(x, h, mode=mode, method=key)
+                           for key in ['fft', 'direct']}
+
+                rtol = {'rtol': 2.0e-3} if dtype in {np.dtype('complex64'),
+                                                     np.dtype('float32')} else {}
+                if isinstance(results['direct'], np.ndarray):
+                    assert_allclose(results['fft'], results['direct'], **rtol)
```
Comment (pv) [inline scipy/signal/tests/test_signaltools.py]: Maybe retain the dtype check also here?

Comment (stsievert) [inline scipy/signal/tests/test_signaltools.py]: What value would that add? This dtype check is less complete than the other checks (this check only looks at same dtypes for 1D; other checks look at all dtype combinations for 1D).

**Comment by Eric89GXL** [comment]
Thanks for looking into this.

FYI your commit history is weird. You might want to go with a fresh branch off of `upstream/master` + `git cherry-pick` for your new commits.

**Comment by stsievert** [comment]
Bug fixed.

I discovered another bug; after calling `fftconvolve`, the output was only rounded when the first argument was an int/uint type. I fixed that bug as well by rounding when either argument is int/uint.

**Comment by Eric89GXL** [comment]
> I discovered another bug; after calling fftconvolve, the output was only rounded when the first argument was an int/uint type. I fixed that bug as well by rounding when either argument is int/uint.

I'm a bit surprised either gets rounded, since I'd assume that the `np.result_type` of the inputs would be used. But if it was that was previously, I guess we'll stick with it.

> The tests are redundant 

Is it worth combining them, then?

**Comment by stsievert** [comment]
> I'm a bit surprised either gets rounded

``` python
In [2]: np.array([1.1, 1.9]).astype(int)
Out[2]: array([1, 1])
```

> Is it worth combining them, then?

Done, and rebased to 3 commits.

**Comment by Eric89GXL** [comment]
> I'm a bit surprised either gets rounded

I mean that if you convolve a double and int, you should get double out (as opposed to talking about rounding vs truncation). Is that the situation you're talking about when you say one argument is int/uint, or something different?

**Comment by stsievert** [comment]
> I mean that if you convolve a double and int, you should get double out

But with `fftconvolve` if convolving two integers results in a float.

Turns out I had the condition for rounding the output incorrect. I've fixed that which resolves #7217.

**Comment by stsievert** [comment]
I've resolved and cleaned the tests. In this, I discovered an issue that resolved #7222 (the issue was an incorrect output rounding condition).

**Comment by Eric89GXL** [comment]
LGTM, +1 for merge. We should probably backport to 0.18, too.

**Comment by Eric89GXL** [comment]
@e-q do you have time to look?

**Comment by Eric89GXL** [comment]
Yeah, 0.19, thanks.

File: scipy/signal/tests/test_signaltools.py [hunk 2]
```diff
@@ -156,41 +156,49 @@ def test_invalid_shapes(self):
...
+        args = [(t1, t2, mode) for t1 in types for t2 in types
+                               for mode in ['valid', 'full', 'same']]
+
+        array_types = {'i': np.random.choice([0, 1], size=n),
+                       'f': np.random.randn(n)}
+        array_types['b'] = array_types['u'] = array_types['i']
+        array_types['c'] = array_types['f'] + 0.5j * array_types['f']
+
+        for t1, t2, mode in args:
+            np.random.seed(42)
```
Comment (e-q) [inline scipy/signal/tests/test_signaltools.py]: This line should be moved up to before `array_types` is created, so the `np.random.choice` calls come from the seed.

File: scipy/signal/tests/test_signaltools.py [hunk 3]
```diff
@@ -156,41 +156,49 @@ def test_invalid_shapes(self):
...
+            if 'bool' in t1 and 'bool' in t2:
+                assert_equal(choose_conv_method(x1, x2), 'direct')
+                continue
+
+            if any([t in {'complex64', 'float32'} for t in [t1, t2]]):
+                kwargs = {'rtol': 8.0e-3, 'atol': 1e-3}
+            elif 'float16' in [t1, t2]:
+                kwargs = {'rtol': 1e-2, 'atol': 1e-5}
+            else:
+                kwargs = {'rtol': 1e-5, 'atol': 1e-5}
```
Comment (e-q) [inline scipy/signal/tests/test_signaltools.py]: Can you add a few comments on how you determined these tolerance levels for the different types?

**Review by e-q — COMMENTED** [review]
Left a few minor comments that I would like to see addressed, but otherwise this looks good. Thanks @stsievert for putting this together!

Something I noticed that could be a candidate for another PR is that `convolve` doesn't appear to check the validity of the `mode` and `method` arguments.

File: scipy/signal/tests/test_signaltools.py (hunk 3 above)
Comment (stsievert) [inline scipy/signal/tests/test_signaltools.py]: Added.

File: scipy/signal/tests/test_signaltools.py (hunk 2 above)
Comment (stsievert) [inline scipy/signal/tests/test_signaltools.py]: Done, thanks for pointing this mistake out.

**Comment by stsievert** [comment]
@e-q's comments have been addressed.

**Comment by e-q** [comment]
Looks good. Squashed and committed in 760a369. Thanks @stsievert!

**Pull request closed without merging**